src/plots/cartesian/axis_defaults.js

```diff
@@ -142,6 +138,11 @@ module.exports = function handleAxisDefaults(containerIn, containerOut, coerce,
         delete containerOut.zerolinewidth;
     }
 
+    // fill in categories
+    containerOut._initialCategories = axType === 'category' ?
+        orderedCategories(letter, containerOut.categoryorder, containerOut.categoryarray, options.data) :
+        [];
+
     return containerOut;
```

Comment:
Thanks @etpinard! It's so useful it would deserve to stick around as a block comment in the code. Sometimes the thoughts and concepts behind coding choices are easily unravelled, but I definitely missed some.